app/flatpak-builtins-search.c

```diff
@@ -156,61 +141,36 @@ compare_by_score (MatchResult *a, MatchResult *b, gpointer user_data)
   return (int) b->score - (int) a->score;
 }
 
-#if !AS_CHECK_VERSION (0, 6, 1)
-/* Roughly copied directly from appstream-glib */
-
-static const gchar *
-as_app_fix_unique_nullable (const gchar *tmp)
-{
-  if (tmp == NULL || tmp[0] == '\0')
-    return "*";
-  return tmp;
-}
-
-static char *
-as_app_get_unique_id (AsApp *app)
-{
-  const gchar *id_str = NULL;
-  const gchar *kind_str = NULL;
-  AsAppKind kind = as_app_get_kind (app);
-
-  if (kind != AS_APP_KIND_UNKNOWN)
-    kind_str = as_app_kind_to_string (kind);
-  id_str = as_app_get_id_no_prefix (app);
-  return g_strdup_printf ("%s/%s",
-                          as_app_fix_unique_nullable (kind_str),
-                          as_app_fix_unique_nullable (id_str));
-}
-
 static gboolean
-as_app_equal (AsApp *app1, AsApp *app2)
+as_app_equal (AsComponent *app1, AsComponent *app2)
 {
   if (app1 == app2)
     return TRUE;
 
-  g_autofree char *app1_id = as_app_get_unique_id (app1);
-  g_autofree char *app2_id = as_app_get_unique_id (app2);
-  return strcmp (app1_id, app2_id) == 0;
+  AsBundle *app1_bundle = as_component_get_bundle (app1, AS_BUNDLE_KIND_FLATPAK);
+  AsBundle *app2_bundle = as_component_get_bundle (app2, AS_BUNDLE_KIND_FLATPAK);
+  const char *app1_ref = as_bundle_get_id (app1_bundle);
+  const char *app2_ref = as_bundle_get_id (app2_bundle);
+  g_autoptr(FlatpakDecomposed) app1_decomposed = flatpak_decomposed_new_from_ref (app1_ref, NULL);
+  g_autoptr(FlatpakDecomposed) app2_decomposed = flatpak_decomposed_new_from_ref (app2_ref, NULL);
```

Comment:
This is going to be extracting and parsing the ref over and over and over. I think you should add a as_app_get_decomposed() helper that does this and stores the result with gobject data on the AsComponent. That will be a lot faster.

Reply:
Ah, i see you added this at the end.